lib/senders/email.js

```diff
@@ -253,97 +267,227 @@ module.exports = function (log, config) {
+    const mailers = {
+      internal: this.mailer,
+      external: this.emailService
+    }
+    const indices = new Map()
+    const emailAddresses = [ message.email ]
+    if (Array.isArray(message.ccEmails)) {
+      emailAddresses.push(...message.ccEmails)
+    }
+
+    return redis.get('config')
+      .catch(() => {})
+      .then(liveConfig => {
+        return emailAddresses.reduce((services, emailAddress) => {
+          if (liveConfig) {
+            const { sendgrid, ses } = liveConfig
+
+            if (sendgrid && isLiveConfigMatch(sendgrid, emailAddress)) {
+              return upsertServicesArrayExternalSendgrid(services, emailAddress)
             }
-          )
-          return d.reject(err)
-        }
-        log.info(
-          {
-            op: 'mailer.send.1',
-            status: status && status.message,
-            id: status && status.messageId,
-            to: emailConfig && emailConfig.to
+
+            if (ses && isLiveConfigMatch(ses, emailAddress)) {
+              return upsertServicesArrayExternalSes(services, emailAddress)
+            }
+          }
+
+          if (config.emailService.forcedEmailAddresses.test(message.email)) {
+            return upsertServicesArrayExternalSes(services, emailAddress)
           }
-        )
 
-        emailUtils.logEmailEventSent(log, message)
+          return upsertServicesArrayInternal(services, emailAddress)
+        }, [])
+      })
+
+    function isLiveConfigMatch (liveConfig, emailAddress) {
+      const { percentage, regex } = liveConfig
+
+      if (percentage >= 0 && percentage < 100) {
+        let sum = 0
 
-        return d.resolve(status)
+        for (const character in emailAddress) {
+          sum += emailAddress.charCodeAt(character)
+        }
+
+        if (sum % 100 >= percentage) {
+          return false
+        }
       }
-    )
-    return d.promise
+
+      if (regex && ! new RegExp(regex).test(emailAddress)) {
+        return false
+      }
+
+      return true
+    }
+
+    function upsertServicesArrayExternalSendgrid (services, emailAddress) {
+      return upsertServicesArray(services, SERVICES.externalSendgrid, emailAddress, {
+        mailer: mailers.external,
+        emailService: 'fxa-email-service',
+        emailSender: 'sendgrid'
+      })
+    }
+
+    function upsertServicesArrayExternalSes (services, emailAddress) {
+      return upsertServicesArray(services, SERVICES.externalSes, emailAddress, {
+        mailer: mailers.external,
+        emailService: 'fxa-email-service',
+        emailSender: 'ses'
+      })
+    }
+
+    function upsertServicesArrayInternal (services, emailAddress) {
+      return upsertServicesArray(services, SERVICES.internal, emailAddress, {
+        mailer: mailers.internal,
+        emailService: 'fxa-auth-server',
+        emailSender: 'ses'
+      })
+    }
+
+    function upsertServicesArray (services, service, emailAddress, data) {
+      if (indices.has(service)) {
```

Comment:
Other than ensuring emails are always added to the correct service, is the order of the returned services array important? I'm wondering why go through the trouble of creating a services array and indices map. 

Would using a single Map where the map key is the service name and the entry is what's below be simpler? MDN says [calling map.entries()](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Map/entries) returns the entries in inserted order, so it should be the same order?

Reply:
I still need to create an array later so that the consuming code can call [`Array.map()`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/map), because `Map.entries()` returns an [`Iterator`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Iterators_and_Generators).

But I've changed it to use [`Map.values()`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Map/values) and [`Array.from()`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/from) instead, since you dislike building the array explicitly. Although fwiw I think that makes for less readable code later on:

```js
return Array.from(emailAddresses.reduce(services, emailAddress) => {
  // ...
}, new Map()).values())
```

 🚲🏡😄

Reply:
It's not that I dislike building an array explicitly, I thought it might be possible to do it using a single data structure rather than 2. If the new way is less readable to you, go back to the old way!